src/SoftRobots.Inverse/component/constraint/BarycentricCenterEffector.cpp

```diff
@@ -38,55 +38,32 @@ using sofa::defaulttype::Vec3Types;
 using sofa::defaulttype::Rigid3Types;
 using sofa::core::RegisterObject ;
 
-
 template<>
-void BarycentricCenterEffector<Rigid3Types>::buildConstraintMatrix(const ConstraintParams* cParams,
-                                                                   DataMatrixDeriv &cMatrix,
-                                                                   unsigned int &cIndex,
-                                                                   const DataVecCoord &x)
+void BarycentricCenterEffector<Rigid3Types>::draw(const VisualParams* vparams)
 {
-    SOFA_UNUSED(cParams);
-    SOFA_UNUSED(x);
-
-    d_constraintIndex.setValue(cIndex);
-    const auto& constraintIndex = sofa::helper::getReadAccessor(d_constraintIndex);
-
-    const unsigned int nbp = m_state->getSize();
-
-    MatrixDeriv& matrix = *cMatrix.beginEdit();
-
-    unsigned int index = 0;
-    const Vec<3, Real> cx(1.0/Real(nbp),0,0), cy(0,1.0/Real(nbp),0), cz(0,0,1.0/Real(nbp));
-    const Vec<3, Real> vZero(0,0,0);
+    if(d_componentState.getValue() != ComponentState::Valid)
+        return;
 
-    if(d_axis.getValue()[0])
-    {
-        MatrixDerivRowIterator rowIterator = matrix.writeLine(constraintIndex+index);
-        for (unsigned int i=0; i<nbp; i++)
-            rowIterator.setCol(i, Deriv(cx,vZero));
-        index++;
-    }
+    if (!d_drawBarycenter.getValue())
+        return;
 
-    if(d_axis.getValue()[1])
-    {
-        MatrixDerivRowIterator rowIterator = matrix.writeLine(constraintIndex+index);
-        for (unsigned int i=0; i<nbp; i++)
-            rowIterator.setCol(i, Deriv(cy,vZero));
-        index++;
-    }
-
-    if(d_axis.getValue()[2])
-    {
-        MatrixDerivRowIterator rowIterator = matrix.writeLine(constraintIndex+index);
-        for (unsigned int i=0; i<nbp; i++)
-            rowIterator.setCol(i, Deriv(cz,vZero));
-        index++;
-    }
-
-    cIndex+=index;
+    computeBarycenter();
+    Coord barycenter = d_barycenter.getValue();
+    vparams->drawTool()->drawFrame(barycenter.getCenter(), barycenter.getOrientation(), sofa::type::Vec3f(5., 5., 5.));
+}
 
-    cMatrix.endEdit();
-    m_nbLines = cIndex - constraintIndex;
+template<>
```

Comment:
you added this right? I did not think about specialization for rigids when I did my suggestions. Have a look if you could use `DataTypes::CPos` to try to factorize the function

Reply:
Well. I've made the changes. The only difference was the normalization of the orientation in the case of a Rigid. 

So I've factorized the code by adding a `setBarycenter()` function which normalize the orientation in the case of a Rigid. It makes sens, even though it hides/disconnects the normalization after the `accumulate`. But I'm fine with it.